Authentication

Xin-Tong opened the pull request:
Closes #11

bboe commented:
> Multiple browsers can concurrently hold permanent cookies.

This requirement is not satisfied. Steps to reproduce:

1) Open firefox, login as a user with "remember password" selected
2) Open chrome, login as the same user with "remember password" selected
3) Close firefox
4) Reopen firefox and observe that the user is not still logged in

Note that if you skipped step 2 the user would still be logged in.

Xin-Tong commented:
@bboe I squashed the modification to remove token from database into https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/284520e7324b7f3bf90895f333a945f795d489ce since it's better to just delete the unused migration.

bboe commented:
Great. We'll see what @d-ray has to say.

d-ray commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    redirect_to root_path,
+                flash: {
+                  warning: "Hi #{@current_user.name}, you've already logged in."
+                } if logged_in?
```
Do you think that an inline `if` after a multi-line method call promotes optimal readability?

Xin-Tong replied:
Good suggestion. I'll address it later.

d-ray commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def create
+    user = User.find_by(email: params[:session][:email].downcase)
```
What will happen if a user is already logged in and tries to re-authenticate?

Xin-Tong replied:
Do you mean the flash of Line7? https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32#discussion-diff-61662695R7

d-ray replied:
Right, line 7 handles this case in the `new` action.  My question is more about what happens for the same situation in the `create` action.

d-ray commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,53 @@
+      log_out!
+      redirect_to login_path
+    elsif User.find_by(id: params[:id])
+      redirect_to root_path, flash: { danger: 'You have already logged out.' }
```
Does having different responses based on whether the id is valid expose any sensitive information?

Xin-Tong replied:
User is searched by user ID instead of user email. So ID here is just a number, shouldn't have security issue.

d-ray commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,25 @@
+      flash[:success] = 'Welcome!'
+      redirect_to login_path
+    else
+      render :new, status: :unprocessable_entity
```
Should there be any indication to the user of what specifically went wrong?

Xin-Tong replied:
👍   I will add a flash message.

Xin-Tong replied:
I am addressing this comment now and realize there are validations according to different input fields (so the according error messages will show as well). Is it necessary to add an extra error message here?

d-ray replied:
Ah, right.  If there are inline form errors you probably don't need a flash message as well.

d-ray commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,25 @@
+  end
+
+  def index
+    @users = User.all
```
What's the purpose of having an `index` action for users?

Xin-Tong replied:
Almost just for double-check at that time. I forgot to delete. Thanks.

d-ray commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,25 @@
+  private
+
+  def user_params
+    params.require(:user).permit(:name, :email, :password)
```
What will happen if `user` isn't included in the params?

Xin-Tong replied:
It will raise ```ActionController::ParameterMissing: param is missing or the value is empty: user``` since we use `require` here. But I don't know how to reproduce it in browser. Is it an issue we need to handle?

d-ray replied:
If you're satisfied with the way Rails handles it, then there's probably no need to customize.

d-ray commented on app/helpers/navbar_helper.rb:
```diff
@@ -1,9 +1,9 @@
 module NavbarHelper
-  def nav_item(text:, path:, is_active:)
+  def nav_item(text:, path:, is_active:, html_options: 'get')
```
Is `html_options` an accurate name given how this parameter is actually used?

Xin-Tong replied:
Maybe rename it as `http_verb` ?

d-ray replied:
That would be a more accurate name.  You could also make `html_options` a hash and read the verb out of it, e.g.

    def nav_item(text:, path:, is_active:, html_options: {})
      http_verb = html_options[:http_verb] || 'get'
      …
    end

but if `http_verb` is the only option you want to support it probably makes more sense to do what you suggested.

Xin-Tong replied:
Sounds good. I will do it. Thanks for elaborating.

d-ray commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -3,6 +3,14 @@
       <%= nav_item(text: 'Index', path: root_path, is_active: current_page?(images_path) || current_page?(root_path)) %>
       <%= nav_item(text: 'New', path: new_image_path, is_active: current_page?(new_image_path)) %>
+      <div class="pull-xs-right">
+        <%= nav_item(text: 'Sign up now!', path: signup_path, is_active: current_page?(signup_path)) %>
```
It seems like the logic for which nav item should be active kind of clutters up this partial a bit.  Are there any cleaner ways to handle this?

Xin-Tong replied:
Will address it.

Xin-Tong replied:
In Line4 we have `is_active: current_page?(images_path) || current_page?(root_path)`, I am not sure how to refactor this part since there are two inputs.

d-ray replied:
How would you refactor it if there was only one input on each line?

Xin-Tong replied:
move the `current_page?(signup_path)` into `nav_item`, so inside `nav_item`, we have something like: ``` is_active = current_page?(path)```

d-ray replied:
Ok, great.  That sounds good; I would probably recommend naming the parameter something that better reflects it's usage though.  It _is_ a path, but in this context it _represents_ the page on which the link should be active.  You might call it `active_page`.  Now, the only difference between what you suggested here and what you need (to handle line 4) is that there can be multiple active pages.  How can you adjust what you have to meet that requirement?

Xin-Tong replied:
Can I do it like this?
```
module NavbarHelper
  def nav_item(text:, active_page:, http_verb: 'get')
    is_active = active_page.any? {|path| current_page?(path)}
    path = active_page[0]
    classes = %w(nav-item)
    classes << 'active' if is_active
    content_tag :li, class: classes do
      link_to text, path, class: 'nav-link', method: http_verb
    end
  end
end
```
`active_page` will be an array.
But I don't like the line of ` path = active_page[0]` that much.

d-ray replied:
> I would probably recommend naming the parameter something that better reflects it's usage though. It is a path, but in this context it represents the page on which the link should be active.

So I somehow overlooked the fact that the path parameter is also being used as the path to the link - `path` is a totally reasonable parameter name.  I agree that `path = active_page[0]` isn't great.  It seems like the issue with it is that it doesn't clearly distinguish between the link path and paths for other pages on which the link should be active.  It would be better to distinguish between these different concerns with a method signature like `def nav_item(text:, page:, other_active_pages: [], http_verb: 'get')`.  However, I don't think you actually need to support this.  What is `root_path` for the application?

Xin-Tong replied:
The `root_path` is `/` which will hit the `images#index` action. We set it in `routes.rb`.

d-ray replied:
Oh, of course.  It's a different path even though it's routed to the same place.  Ok, well you could try what I suggested above.  I think it's an improvement, but perhaps the improvement it offers doesn't justify the effort of making the change.  It's up to you how you want to handle this.

Xin-Tong replied:
As long as there is improvement :) This is the commit https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/713dfa8a7ce46090e27d67ed9c54c3ea3f679def

d-ray commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,7 @@
+  <ul>
+    <li>
+      <%= "Name: #{user.name} Email: #{user.email}" %>
+      </li>
```
The indentation is off here.

Xin-Tong replied:
will address.

d-ray commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,7 @@
+<% @users.each do |user| %>
+  <ul>
```
Do you want the `ul` element inside of the loop?

Xin-Tong replied:
I will pull it out of the loop.

d-ray commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,53 @@
+include SessionsHelper
```
Do you need all of the methods provided by the helper?

d-ray commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,53 @@
+
+  private
+
+  def log_in!(user)
```
This is also defined in the included `SessionsHelper`.

d-ray commented on app/helpers/sessions_helper.rb:
```diff
@@ -0,0 +1,20 @@
+    !current_user.nil?
+  end
+
+  def log_in!(user)
```
Does it make sense to expose a method like this in the view by including it in the helper?

Xin-Tong replied:
I've thought about where to put this method. But I need to use it in the `current_user` . On the other hand, I need to use the `log_in` in controller test. So I chose to put it here and in the test.

d-ray replied:
> I need to use it in the `current_user`

It looks like `current_user` is only called by `logged_in?` (since the controller references `@current_user` directly); `logged_in?` in turn is only called from the view.  As a more general follow up to my original question, does it make sense for the action of logging in a user to happen (or not) based on method calls in the view?  If not, what is the "right" place to take that action?

d-ray replied:
On a totally separate note, what was your reasoning for naming the method with a `!` (i.e. `log_in!` as opposed to `log_in`)?

Xin-Tong replied:
`logged_in?` and `current_user` are used in every view since they're called from nav bar. Does it make sense to move they two into a module called `helpers/common_helpers/login_helper.rb` then include this helper into `application_helper`?

Xin-Tong replied:
Can we discuss about it tomorrow? I'm thinking about use `before_action` but not very sure how to separate.

d-ray commented on app/helpers/sessions_helper.rb:
```diff
@@ -0,0 +1,20 @@
+module SessionsHelper
```
Is there a corresponding test for this helper?

d-ray commented on test/controllers/sessions_controller_test.rb:
```diff
@@ -0,0 +1,94 @@
+  end
+
+  test 'user is logged_in using remember me' do
+    cookies.permanent.signed[:user_id] = @user.id
```
Does this test provide value in its current form?

Xin-Tong replied:
I should `get :new` here. will address it.

d-ray commented on test/models/user_test.rb:
```diff
@@ -0,0 +1,97 @@
+  test 'email addresses should be unique' do
+    duplicate_user = User.new(name: 'Duplicate User', email: 'user@example.com', password: 'something else')
+    duplicate_user.email = @user.email.upcase
+    @user.save
```
Do you need to save the user here?

Xin-Tong replied:
You're right. I originally just want to check if it's valid. I will remove it.

d-ray replied:
I think I'm wrong, actually, but try it and see what happens.

Xin-Tong replied:
I double check that we need the `save` here. Thanks for making me to think twice about it.

d-ray replied:
Sure. 😃  In my original comment I assumed that the user was saved on creation, but after writing it I realized that it wasn't.  I chose to leave the comment just so that you would have to think twice.

Xin-Tong commented:
@d-ray I've addressed all the comments except this one https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32#discussion-diff-61782351R7. I haven't found a optimal way yet. Could you give some input?

Xin-Tong commented:
@d-ray All the comments have been addressed. Please let me know if it's good to squash then merge.

d-ray commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,57 @@
+include ApplicationHelper
+include LoginService
```
Should we be including `LoginService` here?

Xin-Tong replied:
It seems not. I will remove it.

Xin-Tong replied:
Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/363fdd94dbf5447291af4b78023d17910f3a15c4

d-ray commented on app/helpers/navbar_helper.rb:
```diff
@@ -1,9 +1,11 @@
 module NavbarHelper
-  def nav_item(text:, path:, is_active:)
+  def nav_item(text:, path:, http_verb: 'get', additional_active_path: [])
+    all_active_path = additional_active_path << path
```
Using `<<` here will modify the value of `additional_active_path`, which wouldn't necessarily be obvious to someone modifying this code in the future.  It might also be good to pluralize `additional_active_path` since it's an array and thus in theory could be many paths.

rmacklin replied:
👍 Exactly. To expand on that point, generally, you should avoid modifying objects passed as parameters whenever possible. Doing that is prone to causing unintended side effects (that can turn up as corner case bugs), but it can usually be easily avoided. In this case you can use array concatenation instead:
```ruby
all_active_paths = additional_active_paths + [path]
```

Xin-Tong replied:
Great suggestion. I'll change it.

d-ray commented on app/models/user.rb:
```diff
@@ -0,0 +1,21 @@
+class User < ActiveRecord::Base
+  before_save { self.email = email.downcase }
+  validates :name, presence: true, length: { maximum: 50 }
+  validates :email, presence: true, length: { maximum: 255 }, email_format: true, uniqueness: { case_sensitive: false }
```
You don't necessarily have to do this, but you could remove the `case_sensitive` option if you changed `before_save { self.email = email.downcase }` to use a `before_validation` hook instead.  See http://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html for more info on the different callbacks and the order in which they run.

Xin-Tong replied:
Nice. Thank you for the information link.

Xin-Tong replied:
Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/363fdd94dbf5447291af4b78023d17910f3a15c4

d-ray commented on test/controllers/sessions_controller_test.rb:
```diff
@@ -0,0 +1,103 @@
+
+  private
+
+  def log_in!
```
Should this method end in a `!`?  See http://dablog.rubypal.com/2007/8/15/bang-methods-or-danger-will-rubyist for an explanation of ruby conventions around methods ending in `!`.

Xin-Tong replied:
I changed the one in `LoginServe` but forgot this one. I'll address it.

Xin-Tong replied:
Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/363fdd94dbf5447291af4b78023d17910f3a15c4

Xin-Tong replied:
Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/363fdd94dbf5447291af4b78023d17910f3a15c4

d-ray commented on test/flow_tests/users_flow_test.rb:
```diff
@@ -0,0 +1,20 @@
+  test 'user flow' do
+    images_index_page = PageObjects::Images::IndexPage.visit
+
+    new_user_page = images_index_page.add_new_user!
```
In light of https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/files#r63780576, I should probably also mention that we have a page object convention that contradicts the ruby convention:  https://github.com/appfolio/ae_page_objects#conventions.  Thus it's ok that `add_new_user!` here (and other page objects methods) end in `!` even though this is against the ruby convention.  Our convention supersedes the ruby convention in the context of page objects/flow tests, so within this context we know that `!` has a particular meaning.  Outside of the context of page objects/flow tests we should follow the ruby convention whenever possible.  In both cases, following the convention allows us to better communicate our intent to future readers of our code.

Xin-Tong replied:
Yes, I try to follow our convention in all the `!` case in this project ("change current window"). Functionally, after log in, we redirect the user to the index page. So I considered it as a "change window". But as you said before, it against the ruby convention, and the log_in function alone do not necessarily redirect the users. So it makes sense to remove the `!` there. I would prefer to keep `!` here.

d-ray replied:
Sorry, I wasn't suggesting that you remove the `!` here.  Rather, I was pointing out that the ruby convention I mentioned in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/files#r63780576 doesn't apply in this case.  I was trying to avoid confusion, but I may have caused more confusion than if I just hadn't said anything.  Oops 😞

Xin-Tong replied:
I misunderstood your message. Now it makes more sense for both `!` cases. Thanks! 👍

d-ray commented on test/helpers/navbar_helper_test.rb:
```diff
@@ -1,14 +1,20 @@
 require 'test_helper'
 
 class NavbarHelperTest < ActionView::TestCase
+  attr_reader :request
```
I think I may have mentioned some benefits of using `attr_reader` over directly referencing instance variables when we paired.  In my mind, those benefits don't really apply to tests - I think this is more likely to confuse someone by making them think that calling code somewhere will be accessing `request` on this class.  What benefits do you see from using it here?

Xin-Tong replied:
I cannot get access to request variable without this getter but I want to use `request.path = root_path` in the following test to define the current page.

Xin-Tong replied:
Or should I stub current_page?

d-ray replied:
> I cannot get access to request variable without this getter
`attr_reader` should just be wrapping an instance variable of the same name.  Does `@request` not work here?

Xin-Tong replied:
Then it will raise `Minitest::UnexpectedError: RuntimeError: You cannot use helpers that need to determine the current page unless your view context provides a Request object in a #request method`

d-ray replied:
Interesting.  I think stubbing `current_page?` is probably a better solution over trying to set the path on the request.

Xin-Tong replied:
Addressed in https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/5f4c1ee0b0328dda2cf06f22b6900e849a9d266d and https://github.com/AppFolioOnboarding/image-sharer-xin/pull/32/commits/3cebd634bc98dd21af4ef9b5ce7997bcc6d34aae

d-ray commented:
@Xin-Tong I took another look and left a few more comments.  Let me know when you've reviewed and/or addressed them.

Xin-Tong commented:
@d-ray I've addressed all the comments.

d-ray commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,56 @@
+    end
+  end
+
+  def log_out!
```
This method name also violates the ruby convention for `!` methods.

d-ray commented on app/helpers/navbar_helper.rb:
```diff
@@ -1,9 +1,11 @@
-  def nav_item(text:, path:, is_active:)
+  def nav_item(text:, path:, http_verb: 'get', additional_active_paths: [])
+    all_active_paths = additional_active_paths + [path]
+    is_active = all_active_paths.any? {|path| current_page?(path)}
```
Not super important, but you could inline this where it's used on line 6 instead of setting it to a variable.

d-ray commented on config/routes.rb:
```diff
@@ -5,6 +5,13 @@
   end
+  
+  get 'signup' => 'users#new'
+  resources :users, only: [:index, :new, :create]
```
You previously removed the index action from the users controller, but the route is still here.

d-ray commented on test/models/user_test.rb:
```diff
@@ -0,0 +1,97 @@
+    invalid_addresses = %w(user@example,com user_at_foo.org user.name@example. foo@bar_baz.com foo@bar+baz.com)
+    invalid_addresses.each do |invalid_address|
+      @user.email = invalid_address
+      @user.valid?
```
It might be a good to leave a comment here that calling `valid?` is necessary to populate `errors` on `@user`.  Otherwise the purpose of this line may not be obvious to future readers of the code.

d-ray commented:
👍
